File: trunk/gge/gge/math/Vec2.cpp
#include "Vec2.h"
#include "Rot.h"
#include <cassert>
#include <sstream>

namespace gge
{
    const Vec2 Vec2::UNIT_X(1,0);
    const Vec2 Vec2::UNIT_Y(0,1);
    const Vec2 Vec2::ZERO(0,0);
    
    Vec2::Vec2() :
        x(0),
        y(0)
    {
    }
    
    Vec2::Vec2(float _x, float _y) :
        x(_x),
        y(_y)
    {
    }
    
    Vec2::Vec2(const Rot& r, float length)
    {
        x = Math::cos(r.angle) * length;
        y = Math::sin(r.angle) * length;
    }
    
    Vec2::Vec2(const Vec2& v) :
        x(v.x),
        y(v.y)
    {
    }
    
    Vec2& Vec2::operator=(const Vec2& v)
    {
        x = v.x;
        y = v.y;
        return *this;
    }
    
    float& Vec2::operator[](int i)
    {
        assert(i == 0 || i == 1);
        if (i == 0)
            return x;
        else
            return y;
    }
    
    Vec2 Vec2::operator+(const Vec2& v) const
    {
        return Vec2(x + v.x, y + v.y);
    }
    
    Vec2 Vec2::operator-(const Vec2& v) const
    {
        return Vec2(x - v.x, y - v.y);
    }
    
    Vec2 Vec2::operator*(float f) const
    {
        return Vec2(x * f, y * f);
    }
    
    Vec2 Vec2::operator/(float f) const
    {
        assert(f != 0);
        return Vec2(x / f, y / f);
    }
    
    Vec2 Vec2::operator-() const
    {
        return Vec2(-x,-y);
    }
    
    void Vec2::operator+=(const Vec2& v)
    {
        x += v.x;
        y += v.y;
    }
    
    void Vec2::operator-=(const Vec2& v)
    {
        x -= v.x;
        y -= v.y;
    }
    
    void Vec2::operator*=(float f)
    {
        x *= f;
        y *= f;
    }
    
    void Vec2::operator/=(float f)
    {
        assert(f != 0);
        x /= f;
        y /= f;
    }
    
    float Vec2::length() const
    {
        return Math::sqrt(x * x + y * y);
    }
    
    Vec2 Vec2::norm() const
    {
        return *this / length();
    }
    
    float Vec2::dot(const Vec2& v) const
    {
        return x * v.x + y * v.y;
    }
    
    Rot Vec2::getRotTo(const Vec2& v) const
    {
        return (Rot(v) - Rot(*this)).norm();
    }
    
    Vec2 Vec2::lerp(const Vec2& v, float i) const
    {
        return Vec2((v.x - x) * i + x, (v.y - y) * i + y);
    }
    
    bool Vec2::equals(const Vec2& v, float eps) const
    {
        return Math::equal(x,v.x,eps) && Math::equal(y,v.y,eps);
    }
    
    std::string Vec2::toString()
    {
        std::ostringstream outs;
        outs << "Vec2(" << x << ", " << y << ")";
        return outs.str();
    }
    
/*    bool Vec2::pyEquals( const Vec2& v )
    {
        return equals(v,Math::EPSILON);
    }
    
    boost::python::object Vec2::defineClass()
    {
        return boost::python::class_<Vec2>("Vec2", boost::python::init<>() )
            .def(boost::python::init<float,float>())
            .def(boost::python::init<Rot,float>())
            .def_readwrite("x", &Vec2::x)
            .def_readwrite("y", &Vec2::y)
            .def("__repr__", &Vec2::toString)
            .def("__eq__", &Vec2::pyEquals)
            .def(boost::python::self + boost::python::other<Vec2>())
            .def(boost::python::self - boost::python::other<Vec2>())
            .def(boost::python::self * float())
            .def(boost::python::self / float())
            .def("length", &Vec2::length)
            .def("norm", &Vec2::norm)
            .def("dot", &Vec2::dot)
            .def("lerp", &Vec2::lerp)
            .def("getRotTo", &Vec2::getRotTo)
        ;
    }*/
}

File: acce3/core/vec2.py
from util import feq
import math

class Vec2:
  def __init__(self, x, y):
    self.x = float(x)
    self.y = float(y)
    
  def __repr__(self):
    return 'Vec2(%f, %f)' % (self.x, self.y)
    
  def __eq__(self, v):
    return feq(self.x, v.x) and feq(self.y, v.y)
    
  def __getitem__(self, i):
    if i == 0:
      return self.x
    elif i == 1:
      return self.y
    else:
      raise KeyError(i)
      
  def __setitem__(self, i, v):
    if i == 0:
      self.x = v
    elif i == 1:
      self.y = v
    else:
      raise KeyError(i)
    
  def __neg__(self):
    return Vec2(-self.x, -self.y)
    
  def __add__(self, v):
    return Vec2(self.x + v.x, self.y + v.y)
    
  def __sub__(self, v):
    return Vec2(self.x - v.x, self.y - v.y)
    
  def __mul__(self, v):
    if isinstance(v, Vec2):
      return Vec2(self.x * v.x, self.y * v.y)
    else:
      return Vec2(self.x * v, self.y * v)
      
  def __div__(self, v):
    if isinstance(v, Vec2):
      return Vec2(self.x / v.x, self.y / v.y)
    else:
      return Vec2(self.x / v, self.y / v)
      
  def sq_length(self):
    return self.x * self.x + self.y * self.y
    
  def length(self):
    return math.sqrt(self.sq_length())
    
  def angle(self):
    return math.atan2(self.y, self.x)
    
  def norm(self):
    return self / self.length()
    
  def dot(self, v):
    return self.x * v.x + self.y * v.y
    
  def lerp(self, v, i):
    return self + ( v - self ) * i
    
  def rotate(self, r):
    a = self.angle()
    l = self.length()
    return Vec2(l * math.cos(r + a), l * math.sin(r + a))
    
  @staticmethod
  def zero():
    return Vec3(0, 0, 0)
    
  @staticmethod
  def test():
    assert Vec2(1, 2) + Vec2(3, 4) == Vec2(4, 6)
    assert Vec2(1, 2) - Vec2(3, 4) == Vec2(-2, -2)
    assert Vec2(1, 2) * Vec2(3, 4) == Vec2(3, 8)
    assert Vec2(1, 2) * 3 == Vec2(3, 6)
    assert Vec2(10, 15) / Vec2(2, 5) == Vec2(5, 3)
    assert Vec2(10, 15) / 5 == Vec2(2, 3)
    assert feq(Vec2(3, 4).sq_length(), 25)
    assert feq(Vec2(3, 4).length(), 5)
    assert feq(Vec2(0, 1).angle(), math.pi/2)
    assert Vec2(0, -10).norm() == Vec2(0, -1)
    assert feq(Vec2(1, 2).dot(Vec2(3, 4)), 11)
    assert Vec2(0, 0).lerp(Vec2(10, 20), .5) == Vec2(5, 10)
    assert Vec2(1, 0).rotate(math.pi * 1.5) == Vec2(0, -1)
    
Vec2.test()

